Use Case: Blue-Green Deployments (#4263)
hashicorp/waypoint: website/content/docs/intro/use-cases/blue-green-deployment.mdx

---
layout: docs
page_title: Blue-Green Deployments with Waypoint, Nomad and Consul
description: |-
  Waypoint enables advanced deployment workflows like blue-green deployments, orchestrated with Nomad and Consul.
---

# Blue-Green Deployments with Waypoint, Nomad and Consul

> In addition to common methods of deploying applications, Hashicorp Waypoint supports

Reviewer: ```suggestion
In addition to common methods of deploying applications, HashiCorp Waypoint supports
```

the orchestration of advanced deployment workflows, such as blue-green deployments.
A blue-green deployment is where a new version of an application is deployed alongside

> the existing version in an environment. While both versions are running, DNS is

Reviewer: ```suggestion
the existing version in a separate environment. While both versions are running, DNS is
```

I think this is what was intended here?

updated to split traffic flowing to the application between the new and existing
versions of the app. This enables essentially nil downtime for the application
during deployments in most cases, but also enables operators to provide the experience

> of the new version of an application to a certain percentage of users, without

Reviewer: ```suggestion
of the new version of an application to a certain percentage of users without
```

forcing all users to use the newer version. Depending upon feedback and testing
while the new version is live and processing a partial amount of traffic, a decision
is made to either promote or roll back the new version, and DNS is again updated
to route 100% of traffic to all instances running the application.

In the HashiStack, this is made possible with features from the workload orchestration

> tool, Hashicorp Nomad, and the service networking tool, Hashicorp Consul. Nomad

Reviewer: ```suggestion
tool, HashiCorp Nomad, and the service networking tool, HashiCorp Consul. Nomad
```

enables operators to deploy [canary](https://developer.hashicorp.com/nomad/docs/job-specification/update#canary)

> instances of a newer version of an application, without destroying the existing

Reviewer: ```suggestion
instances of a newer version of an application without destroying the existing
```

instances. The canary deployment can be auto-promoted when "healthy", or manually

> promoted by operators. Consul plays a role as well, through service configuration

Reviewer: ```suggestion
promoted by operators. Consul plays a role as well through service configuration
```

entries. The [service resolver](https://developer.hashicorp.com/consul/docs/connect/config-entries/service-resolver)
config entry controls which instances of a service should be matched by downstream
requests made to the service, which can be [filtered](https://developer.hashicorp.com/consul/api-docs/features/filtering)
by service tags. A [service splitter](https://developer.hashicorp.com/consul/docs/connect/config-entries/service-splitter)
config entry splits a configured percentage of traffic to resolvers. These features
of Nomad and Consul work together by the canary instances of a new Nomad job version
registering new instances of a Consul service with [canary tags](https://developer.hashicorp.com/nomad/docs/job-specification/service#canary_tags).
The tags applied to the canary instances of the service match a service resolver.
The service splitter is then updated while both the new and old application instances
are running to split some percentage of traffic between the resolver for the new
and old instances of the service.

To achieve this, a series of commands or API calls would need to be run manually,

> or automated through scripts or some other means; but Waypoint pipelines can orchestrate

Reviewer: We should link to the pipeline overview docs on `Waypoint pipelines`!

this! Waypoint pipelines can execute configured phases of your app's build, deployment,

> and release lifecycle, but also run custom steps, such as modifying Consul config

Reviewer: ```suggestion
and release lifecycle, and run custom steps such as modifying Consul config
```

entries to configure DNS to route traffic to your canary deployment instances.

An example where this is configured is located [here](https://github.com/paladin-devops/hashitalks-deploy-blue-green).
The Waypoint configuration file contains three pipelines, and each one will be
described in the next section.

> 

Reviewer: Before we dive deep into the example, it might be nice to have a small paragraph or bulleted list of required services/tools that need to be configured to follow along. i.e. A Nomad and Consul cluster that Waypoint is connected to, etc. Just so the reader knows what kind of things they might need. Not like a full setup guide, but maybe like a heads up you might want to be familiar with these tools before going forward sort of thing?

Author: Added [here](https://github.com/hashicorp/waypoint/pull/4263/commits/cb9a63bd782a634d22fbc0480c0b03896d1eee34)!

using the Consul provider) is depicted below, where instances of the "app" service
tagged with `blue` are a part of the `blue` subset, and those tagged with `green` are
a part of the `green` subset:

```terraform
resource "consul_config_entry" "app_service_resolver" {
  name = "app"
  kind = "service-resolver"

  config_json = jsonencode({
    Kind          = "service-resolver"
    Name          = "app"
    DefaultSubset = "blue"
    Subsets = {
      blue = {
        Filter = "blue in Service.Tags"
      }
      green = {
        Filter = "green in Service.Tags"
      }
    }
  })
}
```

Since a Nomad job is being deployed in this pipeline, the tags on the service instances
are derived from the Nomad jobspec template. Specifically, this is configured in the
`service` stanza of the Nomad job. The `tags` are applied to the Nomad allocations
running the current active version of the Nomad job, and `canary_tags` are applied
to the canary Nomad allocations rolled out during a deployment.

```hcl
    service {
      name        = "app"
      port        = "http"
      tags        = [ "blue" ]
      canary_tags = [ "green" ]

>     }

Reviewer: We should left align this and remove the extra whitespace since it's all on its own!

```

The exact number of canary Nomad allocations to be deployed, with the `canary_tags`
from the `service` stanza, is configured in the `update` stanza's `canary` field:

```hcl
    update {
      max_parallel = 1
      canary       = 1
      auto_revert  = true
      auto_promote = false
      health_check = "task_states"

>     }

Reviewer: Left align this I think and it will probably look nicer in the rendered code block!
